pkg/util/lsof/lsof_linux.go

```diff
@@ -0,0 +1,383 @@
+		// knowing whether the file is memory mapped or program text would require knowing permissions
+		// which aren't parsed by gopsutil, so we just assume it's memory mapped
+		fd = "mem"
+	} else if ty == "DIR" {
+		if path == "/" {
+			fd = "rtd"
+		} else if cwd != "" && path == cwd {
+			fd = "cwd"
+		}
+	}
+	return fd
+}
+
+func (ofl *openFilesLister) fdMetadata() (Files, error) {
+	openFDs, err := ofl.proc.FileDescriptors()
+	if err != nil {
+		return nil, err
+	}
+
+	var files Files
+	for _, openFD := range openFDs {
+		file, ok := ofl.fdStat(openFD)
+		if ok {
+			files = append(files, file)
+		}
+	}
+
+	return files, nil
+}
+
+func (ofl *openFilesLister) fdStat(fd uintptr) (File, bool) {
+	var err error
+	file := File{
+		Fd: fmt.Sprintf("%d", fd),
+	}
+
+	fdLinkPath := fmt.Sprintf("%s/fd/%d", ofl.procPIDPath(), fd)
+
+	var inode uint64
+	if file.Type, file.OpenPerm, _, _ = fileStats(ofl.lstat, fdLinkPath); file.Type == "" {
+		return File{}, false
+	}
+
+	var ok bool
+	// remove some unnecessary information from permissions string
+	// if the expectations are not met, the permissions are left as is
+
+	// file descriptors always have no sticky bit, setuid, setgid
+	file.OpenPerm = strings.TrimPrefix(file.OpenPerm, "-")
+	// file descriptors always have no permission for group and others
+	file.OpenPerm, ok = strings.CutSuffix(file.OpenPerm, "------")
+	if ok {
+		// file descriptors always have execute permission
+		file.OpenPerm = strings.TrimSuffix(file.OpenPerm, "x")
+	}
+
+	if file.Type, file.FilePerm, file.Size, inode = fileStats(ofl.stat, fdLinkPath); file.Type == "" {
+		return File{}, false
+	}
+
+	if info, ok := ofl.socketInfo[inode]; ok {
+		file.Name = info.Description
+		file.FilePerm = info.State
+		file.Type = info.Protocol
+	} else {
+		file.Name, err = ofl.readlink(fdLinkPath)
+		if err != nil {
+			return File{}, false
+		}
+	}
+
+	return file, true
+}
+
+// TCP state codes, from the Linux kernel
+// https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/net/tcp_states.h
+//
+// The same codes are used for UDP, but only ESTABLISHED and CLOSE are valid
+var tcpStates = map[uint64]string{
+	1:  "ESTABLISHED",
+	2:  "SYN_SENT",
+	3:  "SYN_RECV",
+	4:  "FIN_WAIT1",
+	5:  "FIN_WAIT2",
+	6:  "TIME_WAIT",
+	7:  "CLOSE",
+	8:  "CLOSE_WAIT",
+	9:  "LAST_ACK",
+	10: "LISTEN",
+	11: "CLOSING",
+	12: "NEW_SYN_RECV",
+	13: "BOUND_INACTIVE",
+}
+
+func stateStr(state uint64) string {
+	if s, ok := tcpStates[state]; ok {
+		return s
+	}
+	return fmt.Sprintf("UNKNOWN(%d)", state)
+}
+
+// readSocketInfo reads the socket information from /proc/<pid>/net/{tcp,tcp6,udp,udp6,unix}
+// returns a map of inode to socketInfo
+// see https://www.kernel.org/doc/Documentation/networking/proc_net_tcp.txt
+func readSocketInfo(procPIDPath string) map[uint64]socketInfo {
+	si := make(map[uint64]socketInfo)
+
+	fs, err := procfs.NewFS(procPIDPath)
+	if err != nil {
+		log.Debugf("Failed to read %s: %s", procPIDPath, err)
```

Comment:
rather than empty socket inform should we return an error? We could use that error to populate the flare file with information about the issue when trying to list open files

Reply:
Even if we can't get socket info, we can still get a lot of other information, so we might as well keep collecting everything else

Logs emitted while generating the flare will be included in the `flare_creation.txt` file so this error is definitely still visible

Since the error can be seen by checking agent logs and/or `flare_creation.txt`, I don't think also having it in the created file is useful (if we have a lot of errors I'm worried it would rather pollute the output...), what do you think ?

Reply:
I had no idea that all the logs are aggregated in the `flare_creation.txt`, in that case my suggestion no longer makes sense 😄 

Thanks for the information about `flare_creation.txt` file, very useful 🎉

Reply:
Actually it's not all logs, just the logs done using `FlareBuilder.Logf`, so the log you commented on won't be in the file, my bad 😕